vault/diagnose/raft_checks.go

```diff
@@ -0,0 +1,102 @@
+package diagnose
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/hashicorp/vault/physical/raft"
+)
+
+// RaftFilePermsChecks inputs the raft file path and outputs:
+// 1. A bool, which dictates whether Owner has rw permissions
+// 2. A list of errors
+func RaftFilePermsChecks(ctx context.Context, path string) {
+	owner := "owner"
+	group := "group"
+	other := "other"
+	var overPermissions []string
+
+	// Note: Stat does not return information about the symlink itself, in the case where we are dealing with one.
+	info, err := os.Stat(path)
+	if err != nil {
+		SpotError(ctx, "raft file permission checks", fmt.Errorf("error computing file permissions: %w", err))
+	}
+
+	mode := info.Mode()
+
+	canRead := false
+	canWrite := false
+	for i := 1; i < 4; i++ {
+		perm := string(mode.String()[i])
+		fmt.Println("hi")
+		fmt.Printf("%s", perm)
+		if perm != "w" && perm != "r" && perm != "-" {
+			overPermissions = append(overPermissions, owner+":"+perm)
+		} else if perm == "w" {
+			canWrite = true
+		} else if perm == "r" {
+			canRead = true
+		}
+	}
+
+	for i := 4; i < 7; i++ {
+		perm := string(mode.String()[i])
+		fmt.Printf("%s", perm)
+		if perm != "-" {
+			overPermissions = append(overPermissions, group+":"+perm)
+		}
+	}
+
+	for i := 7; i < 10; i++ {
+		perm := string(mode.String()[i])
+		fmt.Printf("%s", perm)
+		if perm != "-" {
+			overPermissions = append(overPermissions, other+":"+perm)
+		}
+	}
+
+	okFlag := true
+	if mode&os.ModeSymlink != 0 {
+		okFlag = false
+		SpotWarn(ctx, "raft file permission checks", "raft storage backend file is a symlink")
+	}
+	if len(overPermissions) > 0 {
+		okFlag = false
+		SpotWarn(ctx, "raft file permission checks", fmt.Sprintf("too many permissions -- %s", strings.Join(overPermissions, ",")))
+	}
+	if !canRead || !canWrite {
+		okFlag = false
+		SpotError(ctx, "raft file permission checks", fmt.Errorf("error: not enough permissions -- owner read: %v, owner write: %v", canRead, canWrite))
+	}
+	if okFlag {
+		SpotOk(ctx, "raft file permission checks", "correct permissions to raft storage backend file path")
+	}
```

Comment:
One other thing I'd like if you can pull it off is to check whether the target files are owned by root and only owner has read/write perms.  That would imply Vault has to run as root which we should warn about.  Something like "raft backend files owned by root and only accessible as root" with an Advise of "This prevents Vault from running as a non-privileged user".

Reply:
That sounds valuable for sure, I can look into it!